Create a new Studio Component named RecommendedNextAction

Jondyr opened the pull request:
## Description

* Added new component to @studio/components `RecommendedNextAction`
* Provides props to 
    * Edit labels on buttons, title- and description texts.
    * Hide save button and events for save and skip button.
* Added a new Storyboard entry with docs and example use case
* Context `StudioRecommendedNextActionContext` to track recommendation suggestions

https://github.com/user-attachments/assets/9c6340fe-6c3d-4590-b5b7-effa304b1db0

### Changes to existing code:
* Moved function to validate a new task ID to a new hook for reuse. ( `EditTaskId.tsx` -> `useValidateBpmnTaskId.tsx`)

## Related Issue(s)

- #12633 
- #12632 

## Verification

- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)

## Documentation

- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)
- [x] Storybook documentation/entry added

Jondyr commented on frontend/packages/process-editor/src/hooks/useBpmnEditor.ts:
```diff
@@ -46,10 +46,13 @@ export const useBpmnEditor = (): UseBpmnViewerResult => {
 
   const updateBpmnDetailsByTaskEvent = useCallback(
-    (taskEvent: TaskEvent) => {
+    (taskEvent: TaskEvent, gfx?: SVGElement, originalEvent?: Event, justAdded: boolean = false) => {
```
bpmnjs sends three parametersØ
https://github.com/bpmn-io/diagram-js/blob/49af799e381f8a5b419f9030b968d7201f949920/lib/features/interaction-events/InteractionEvents.js#L62

Jondyr commented:
@Ildest @Annikenkbrathen 
Added you as reviewers in case you have any comments on the implementation. Added a video to the PR description above that shows the component in use.

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.tsx:
```diff
@@ -0,0 +1,46 @@
+    <StudioCard>
+      <StudioCard.Header>{title}</StudioCard.Header>
+      <StudioCard.Content>
+        <p className={classes.description}>{description}</p>
```
We use <StudioParagraph/>, a wrapper around @digidr/designsystemet paragraph, instead of native tags. This ensures we get the correct styling, sizing, and spacing. :)

framitdavid commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/ConfigContent.tsx:
```diff
@@ -14,6 +14,7 @@ import { EditPolicy } from './EditPolicy';
 import { EditDataTypesToSign } from '../EditDataTypesToSign';
 import { EditUniqueFromSignaturesInDataTypes } from '../EditUniqueFromSignaturesInDataTypes';
 import { StudioModeler } from '../../../utils/bpmnModeler/StudioModeler';
+import { NewNameRecommendation } from '@altinn/process-editor/components/ConfigPanel/ConfigContent/EditTaskRecommendedItems/NewNameRecommendation';
```
Would it be simpler to do this? 😊 

```suggestion
import { NewNameRecommendation } from './EditTaskRecommendedItems/NewNameRecommendation';
```

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.tsx:
```diff
@@ -0,0 +1,46 @@
+        <p className={classes.description}>{description}</p>
+        {children}
+        <div className={classes.buttonGroup}>
+          {!hideSaveButton && (
```
Should we always display the save button and provide a validation error if the user clicks it without entering a value in the input field? 🤔This might be a question for @Annikenkbrathen as well. 😊

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.tsx:
```diff
@@ -0,0 +1,46 @@
+  title: string;
+  description: string;
+  hideSaveButton?: boolean;
+  children?: React.ReactNode;
```
Should we make this prop mandatory instead? Is there any reason to use this component without children? 😊

framitdavid replied:
I see your good attation behind this! 👏  But I think that would be out-of-scope for this component. We do have a meeting next week about alerts and confirms, to figure out how to solve approve/decline challenge. 😊

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.tsx:
```diff
@@ -0,0 +1,46 @@
+import classes from './StudioRecommendedNextAction.module.css';
+import { StudioCard } from '../StudioCard';
+
+export interface StudioRecommendedNextActionProps {
```
I think we mostly are using `type` instead of `interface`. 😊 
```suggestion
export type StudioRecommendedNextActionProps {
```

framitdavid commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/EditTaskRecommendedItems/NewNameRecommendation.tsx:
```diff
@@ -0,0 +1,70 @@
+import React, { useState } from 'react';
+import { useTranslation } from 'react-i18next';
+
+interface NewNameRecommendationProps {}
+
+export const NewNameRecommendation = ({}: NewNameRecommendationProps): React.ReactElement => {
```
No need to specify props type for components without props. 😊 
```suggestion
export const NewNameRecommendation = (): React.ReactElement => {
```

framitdavid commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/EditTaskRecommendedItems/NewNameRecommendation.tsx:
```diff
@@ -0,0 +1,70 @@
+
+      metadata: {
+        ...bpmnDetails.metadata,
+        justAdded: false,
```
I'm just thinking, but from my point of view, I think `justAdded` is quite a loose name. It's difficult to understand what is used for. 

Another thing I think is that adding this data to the metadata on the `BpmnDetails` might not be that reusable. Would this data be sent over the HTTP to the backend as well?🤔 

What about adding a Context which holds an array of information about the items that should show NextRecommendedAction? Then we can also use the same approach for NextRecommendedAction in other places within Studio where we do not have BpmnDetails. 😊

Jondyr replied:
> Another thing I think is that adding this data to the metadata on the BpmnDetails might not be that reusable. Would this data be sent over the HTTP to the backend as well?🤔

As far as i know, only the element property for this type is sent to the backend 🤔 

> What about adding a Context which holds an array of information about the items that should show NextRecommendedAction? Then we can also use the same approach for NextRecommendedAction in other places within Studio where we do not have BpmnDetails. 😊

I'm curious about this, if i understand your suggestion correctly it would be something like this?
```typescript
RecommendedActionStack = [
    { type: "ADDED_BPMN_TYPE",              id: "activity_AbCd1234" },
    { type: "POSSIBLE_UPDATE_TO_NEW_TYPE",  id: "formItem_123" }
]
 ```
And each component using this context would fetch relevant entries from this array? I'm down to talk about this in a huddle if you want to discuss more in detail.

framitdavid replied:
Yes, I was thinking about something like that. Do we need an entire object with type and id? I do think we need to care about the order of the items, I believe it's easier to check if the item exists within the object. If it does exits then show the next recommended action.


### The Context
```typescript
// I made it an object here to enable a hashmap lookup instead of looping through an array. This is more efficient/performant.
const actions: { [key: string]: string } = { "activity_AbCd1234" , "formItem_123"};

const shouldDisplayRecAction = (actionId: string): boolean => {
   return actions[actionId];
}

const removeAction = (actionId: string): void => {
   delete actions[actionId];
}
```


### Component that might show next recommended action
```tsx
// Within the component that might show NextRecommendedAction

const { shouldDisplayNextRecAction, removeAction } = useNextRecommendedActionContext();

// This data will be dynamic based on props/data.
const componentId = "formItem_123";

const onRecommendedActionClosed = (): void => {
   removeAction(componentId);
}

if (shouldDisplayNextRecAction(componentId) {
  return <NextRecommendedAction onSkip={onRecommendedActionClosed} onSave={onRecommendedActionClosed }/>
}

return (<div>The other code to render</div>);
```


I’m happy to join a pair-programming session to discuss this further. The code above needs adjustments before it will compile and run, but it was written just to explain my thoughts.

framitdavid commented on frontend/packages/process-editor/src/types/BpmnDetails.ts:
```diff
@@ -8,4 +8,8 @@ export interface BpmnDetails {
   element?: Element;
+
+  metadata?: {
+    justAdded?: boolean;
```
Ref. previous comment. Let's try to create a generic and reusable concept to know when to show RecommendedNextAction. 😊

framitdavid reviewed (changes requested):
Great! I have a few thoughts and suggestions on creating a more reusable concept for tracking when to show RecommendedNextAction. 😊 I'm up for pair-programming if you are!

Annikenkbrathen commented:
I think I might have arrived a bit late to the party here. I'm not sure if you still have any questions for UX regarding this. I think it looked good in the video. Good job!  The header seemed a bit large compared to the rest of the content in the panel, but you used the smallest header size?

In the sketches, I created a dark background in the panel to indicate that it's a modal overlaying the other content. Why wasn't that included?

Jondyr commented:
@Annikenkbrathen 
> The header seemed a bit large compared to the rest of the content in the panel, but you used the smallest header size?

I've updated the component to use a Heading component with size xs 👍 

> In the sketches, I created a dark background in the panel to indicate that it's a modal overlaying the other content. Why wasn't that included?

I missed this detail in the design, I've updated the background color of the config element to simulate a modal, as we do not have capability to contain a modal inside a given space. Is this a sufficient solution, or is this intended to be a modal with content displayed behind a translucent background?

Updated view:
![image](https://github.com/user-attachments/assets/0fcdb67a-5c05-4a30-9c1a-d495834688e8)

Annikenkbrathen commented:
> @Annikenkbrathen
> 
> > The header seemed a bit large compared to the rest of the content in the panel, but you used the smallest header size?
> 
> I've updated the component to use a Heading component with size xs 👍
> 
> > In the sketches, I created a dark background in the panel to indicate that it's a modal overlaying the other content. Why wasn't that included?
> 
> I missed this detail in the design, I've updated the background color of the config element to simulate a modal, as we do not have capability to contain a modal inside a given space. Is this a sufficient solution, or is this intended to be a modal with content displayed behind a translucent background?
> 
> Updated view: ![image](https://private-user-images.githubusercontent.com/1525466/357067966-0fcdb67a-5c05-4a30-9c1a-d495834688e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NzA5NDgsIm5iZiI6MTcyMzQ3MDY0OCwicGF0aCI6Ii8xNTI1NDY2LzM1NzA2Nzk2Ni0wZmNkYjY3YS01YzA1LTRhMzAtOWMxYS1kNDk1ODM0Njg4ZTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTJUMTM1MDQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjBkNjUwMjg4ZTkyNjEyMTQxMmQ0Y2UwNjBhZTI2ZDY3NGQxMDlmYzI2NDNmZjkzZGQ1MTcxOGY2YThmMTI3MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2OKVYWAeISFFU_QDt6EKbH0_Zj-d64Yj_ZaXwL1NjUQ)

LGTM :D

Jondyr commented:
@framitdavid 
Summary of [changes done since your review](https://github.com/Altinn/altinn-studio/pull/13296/files/34daa8a09bab5007da5556b55ca206f85a31c525..b2ce85d15d0c1d80d5e5c454fe9d3637e24f4f73):
- Use `paragraph` studio wrapper, and add a new wrapper for `heading` (just a placeholder/redirect)
- Add actions to storyboard entry. Just a very simple implementation
- Made `children` non-optional for the component
- Created new context `StudioRecommendedNextActionContext`
  Changed adding of bpmn task to register the task id in this context, and the ConfigContent component queries the context with the task ID to check if the recommended rename action should be displayed

framitdavid commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/EditTaskRecommendedItems/NewNameRecommendation.tsx:
```diff
@@ -0,0 +1,62 @@
+
+  const saveNewName = () => {
+    removeAction(bpmnDetails.element.id);
+    modeling.updateProperties(bpmnDetails.element, {
+      id: newName,
+    });
+    setBpmnDetails({
+      ...bpmnDetails,
+      id: newName,
+    });
+  };
```
Ref. previous comment, we do have abstraction to simplify the usage of the modelling tools from bpmnjs package.

```suggestion
    studioModeler.updateElementProperties({ id: newName });
    setBpmnDetails({
      ...bpmnDetails,
      id: newName,
    });
  };
```

framitdavid commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/EditTaskRecommendedItems/NewNameRecommendation.test.tsx:
```diff
@@ -0,0 +1,84 @@
+      setBpmnDetails: setBpmnDetails,
+      modelerRef: { current: { get: () => ({ updateProperties: jest.fn() }) } },
+    });
+    (useStudioRecommendedNextActionContext as jest.Mock).mockReturnValue({
+      removeAction: jest.fn(),
+      addAction: jest.fn(),
+      shouldDisplayAction: jest.fn(),
+    });
```
What about creating a renderNewCommendedItems method which wraps the context? Then I do not think we need this mock at all? See f.ex `renderConfigContent` within `ConfigContent.tsx`.

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/context/StudioRecommendedNextActionContext.tsx:
```diff
@@ -0,0 +1,52 @@
+import React, { createContext, useMemo, useState } from 'react';
```
Great job with the context! This is exactly what I meant by developing a reusable pattern. Well done! 👏

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.stories.tsx:
```diff
@@ -0,0 +1,70 @@
+
+type ExampleUseCase = StoryFn<typeof StudioRecommendedNextAction>;
+export const ExampleUseCase: ExampleUseCase = (): React.ReactElement => {
+  const [name, setName] = useState('');
+  const onSave = () => {};
+
+  const onSkip = () => {};
```
You can use the fn function from Storybook to capture events, which will then be displayed in the Action tab when the onClick event is triggered.
```suggestion
   import { fn } from '@storybook/test';
   
   const [name, setName] = useState('');
  const onSave = fn() ;
  const onSkip = fn();
```

Jondyr replied:
I see i've implemented it only for the main preview story and not the example case, good catch 👍

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.module.css:
```diff
@@ -0,0 +1,11 @@
+  display: flex;
+  justify-content: flex-end;
+  align-items: flex-start;
+  gap: 12.5px;
```
Maybe we can use a spacing token from the design system instead? 😊 
```suggestion
  gap: var(--fds-spacing-3)
```

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.test.tsx:
```diff
@@ -0,0 +1,85 @@
+import userEvent from '@testing-library/user-event';
+
+describe('StudioRecommendedNextAction', () => {
+  const user = userEvent.setup();
```
It's recommended by Testing Library to define user within each `it` or `test` instead. 😊 Ref: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.test.tsx:
```diff
@@ -0,0 +1,85 @@
+    );
+
+    await user.click(screen.getByText('Save'));
+    expect(onSave).toHaveBeenCalled();
```
should we change the test to ensure it has only been invoked once? 😊 
```suggestion
    expect(onSave).toHaveBeenCalledTimes(1);
```

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.test.tsx:
```diff
@@ -0,0 +1,85 @@
+    );
+
+    await user.click(screen.getByText('Skip'));
+    expect(onSkip).toHaveBeenCalled();
```
Same as previus comment with `toHaveBeenCalledTimes(1)`. :)

framitdavid commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/StudioRecommendedNextAction.test.tsx:
```diff
@@ -0,0 +1,85 @@
+      </StudioRecommendedNextAction>,
+    );
+
+    expect(screen.queryByText('Save')).toBeNull();
```
Could we write the selector like this? 😊 
```suggestion
    expect(screen.queryByRole('button', { name: 'Save' })).not.toBeInTheDocument();
```

framitdavid reviewed (changes requested):
Nice and clean code, made it easy to do review, well done to you!! 👏 🤩 I have some comments, considerations and questions to be answered before we merge this. 😊

Jondyr commented on frontend/libs/studio-components/src/components/StudioRecommendedNextAction/context/StudioRecommendedNextActionContext.tsx:
```diff
@@ -0,0 +1,52 @@
+      addAction,
+      removeAction,
+    }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
I might be wrong in my reasoning behind this memoization, but the idea was that unless `actions` does not change, we do not need to recreate the context value. As this will cause rerenders of all components depending on this context.
But i did some more testing and the memoization seems to have zero effect on rerenders in this case. 
So i'll remove it unless you have any arguments for it's reason to stay?

Jondyr commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/ConfigContent.module.css:
```diff
@@ -4,6 +4,15 @@
   gap: var(--fds-spacing-2);
 }
 
+.grayedBackground {
+  background-color: #bfc4cb;
+}
+
+.grayedBackgroundContent {
```
I like the backdrop name, good suggestion 👍

Jondyr replied:
@framitdavid I'm not sure where we set the fds variables, i'm assuming they come from figma somehow?

framitdavid replied:
You can find the theme palette here: https://www.designsystemet.no/grunnleggende/designelementer/farger 😊

framitdavid replied:
Nice! 👏

Jondyr replied:
I did not find an exact match, but one very similar:
![image](https://github.com/user-attachments/assets/26e96d82-d0d6-40b5-8da6-9dba1513d03c)
left side is `--fds-semantic-border-divider-default`, right side is previous value. Not sure if using variables designed for a different use case (border divider color) is preferred? I've pushed a commit with this change.

Jondyr commented on frontend/packages/process-editor/src/components/ConfigPanel/ConfigContent/ConfigContent.tsx:
```diff
@@ -48,46 +55,57 @@ export const ConfigContent = (): React.ReactElement => {
+          <>
+            <EditTaskId />
+            <StudioDisplayTile
+              label={t('process_editor.configuration_panel_name_label')}
```
I've pushed a commit related to this, I think this looks a lot better. The logic is easier to follow in my opinion, and we don't have to conditionally apply the css classes, as you state. 👍

framitdavid reviewed (changes requested):
Great work! Just a few small comments on some of the tests. 😊

framitdavid reviewed (approved):
👏

ErlingHauan reviewed (approved):
### Testing
Nice feature and implementation 👏 

@Annikenkbrathen: Nitpicking, but I noticed that the font sizing is different between the normal view and RecommendedNextAction. It's not crucial by any means, but it would maybe look a little bit more cohesive if the sizing was the same?

@Jondyr: It is not possible to save the new ID by pressing enter. If it is not a quick fix, maybe it can be a separate issue.

But overall, it looks and feels like a solid component 😄

Annikenkbrathen commented:
> ### Testing
> Nice feature and implementation 👏
> 
> @Annikenkbrathen: Nitpicking, but I noticed that the font sizing is different between the normal view and RecommendedNextAction. It's not crucial by any means, but it would maybe look a little bit more cohesive if the sizing was the same?

Agreed, well observed. I still think the texts here are one size too large. Are you using small? It should be 15px. And the header should be x-small, 21px.
<img width="664" alt="Skjermbilde 2024-08-21 kl  09 35 28" src="https://github.com/user-attachments/assets/e49eef67-e08c-4bcf-a237-679c52ff2e87">
<img width="670" alt="Skjermbilde 2024-08-21 kl  09 35 39" src="https://github.com/user-attachments/assets/5fa3a014-9ea8-43b8-925c-af19663ce18e">

Is there a reason you moved away from the background color under the component?

I also notice that this field no longer takes up the full width. Is this part of the same issue, or does it belong elsewhere
<img width="440" alt="Skjermbilde 2024-08-21 kl  09 24 49" src="https://github.com/user-attachments/assets/38b9a4bb-c91a-4c1e-9e21-2b27c66f793d">

Jondyr commented:
@ErlingHauan:
> @Jondyr: It is not possible to save the new ID by pressing enter. If it is not a quick fix, maybe it can be a separate issue.

This is now implemented. 👍 

@Annikenkbrathen:
> Agreed, well observed. I still think the texts here are one size too large. Are you using small? It should be 15px. And the header should be x-small, 21px.

This was a misunderstanding on my part, the paragraph is set to small now. The header is x-small already and should be 21px (for some reason it's actually 20.96 px).

> Is there a reason you moved away from the background color under the component?

I am confused how this happened, but this was not intended and most likely happened when i commited the code changes for the changes [discussed here](https://github.com/Altinn/altinn-studio/pull/13296#discussion_r1716524152). This is now fixed.

> I also notice that this field no longer takes up the full width. Is this part of the same issue, or does it belong elsewhere

Good catch, this issue was introduced when refactoring the component and a css class was lost. This is now fixed as well.

ErlingHauan commented:
Very nice @Jondyr!

One thing I noticed in the code is the use of fireEvent in some of the tests. Can you replace them with userEvent? I think it technically does not matter which one we use, but we are using userEvent elsewhere in the code base, and should be consistent with it 😊

Jondyr commented:
I noticed the header for the whole panel in the Figma sketch is 24px, while on master/staging it's 18px. That might contribute to the feeling that the rest of the fonts are the wrong size? 🤔 I can possibly change this to match the figma sketch size, if preferred? @Annikenkbrathen 

Figma:
![image](https://github.com/user-attachments/assets/c589aaea-cfd1-4146-83b7-666d64971a9b)
Staging:
![image](https://github.com/user-attachments/assets/61d26454-9217-438b-859e-d31fae1774d4)

Also the text looks slightly off-centered compared to the icon on staging to me, maybe this can be solved at the same time. Unless this is part of another task?

Jondyr commented:
> Very nice @Jondyr!
> 
> One thing I noticed in the code is the use of fireEvent in some of the tests. Can you replace them with userEvent? I think it technically does not matter which one we use, but we are using userEvent elsewhere in the code base, and should be consistent with it 😊

I agree, userEvent is what i would have preferred to use. But it seems userEvent does not support forms, as i got `not implemented` errors when i changed the save button to `type='submit'`. I did not look into this too much, so i can investigate a bit more after daily.
